src/v/archival/tests/async_data_uploader_test.cc

```diff
@@ -0,0 +1,967 @@
+        auto closer = ss::defer([&] {
+            cc_stream.close().get();
+            file_out_s.close().get();
+        });
+
+        vlog(
+          test_log.info,
+          "Expected size: {}, \n{}",
+          buf.size_bytes(),
+          buf.hexdump(1024));
+        return buf;
+    }
+
+    model::timestamp _produce_timestamp{0};
+};
+
+inline upload_reconciliation_result
+consume_segment_and_compute_metadata(const iobuf& s) {
+    vlog(
+      test_log.debug,
+      "consume_segment_and_compute_metadata {} bytes",
+      s.size_bytes());
+    upload_reconciliation_result meta = {};
+    meta.size_bytes = 0;
+    iobuf out;
+    auto dst = make_iobuf_ref_output_stream(out);
+    auto src = make_iobuf_input_stream(s.copy());
+    auto pred = [&meta](model::record_batch_header& hdr) {
+        if (meta.offsets.base == model::offset()) {
+            meta.offsets.base = hdr.base_offset;
+        }
+        meta.offsets.last = hdr.last_offset();
+        meta.size_bytes += hdr.size_bytes;
+        vlog(
+          test_log.debug,
+          "{} bytes consumed, O: {}-{}, T: {}-{}",
+          meta.size_bytes,
+          hdr.base_offset,
+          hdr.last_offset(),
+          hdr.first_timestamp,
+          hdr.max_timestamp);
+        return storage::batch_consumer::consume_result::accept_batch;
+    };
+    ss::abort_source as;
+    auto len = storage::transform_stream(
+                 std::move(src), std::move(dst), pred, as)
+                 .get();
+    BOOST_REQUIRE(len.has_value());
+    BOOST_REQUIRE_EQUAL(len.value(), meta.size_bytes);
+    return meta;
+}
+
+struct random_records_generator {
+    std::vector<tests::kv_t> operator()() {
+        std::vector<tests::kv_t> batch;
+        for (size_t i = 0; i < records_per_batch; i++) {
+            auto key = get_next_key();
+            vlog(test_log.info, "Used key {}", key);
+            auto record = random_generators::gen_alphanum_string(record_size);
+            batch.emplace_back(std::move(key), std::move(record));
+        }
+        return batch;
+    }
+    size_t key_size{10};
+    size_t record_size{1000};
+    size_t records_per_batch{10};
+    std::optional<int> key_space_size;
+
+private:
+    ss::sstring get_next_key() {
+        if (key_space_size.has_value() && keys.empty()) {
+            // Initialize keys list
+            for (int i = 0; i < key_space_size.value(); i++) {
+                auto key = random_generators::gen_alphanum_string(key_size);
+                vlog(test_log.info, "Generated key {}", key);
+                keys.emplace_back(std::move(key));
+            }
+        }
+        if (key_space_size.has_value()) {
+            auto ix = random_generators::get_int(0, (int)keys.size() - 1);
+            return keys.at(ix);
+        }
+        return random_generators::gen_alphanum_string(key_size);
+    }
+    std::vector<ss::sstring> keys;
+};
+
+void dump_to_disk(iobuf buf, ss::sstring fname) {
```

Comment:
nit: this looks unused? (or perhaps there for debugging?)

Reply:
it was used by debug logging, might be useful in the future